kafka-rest/src/main/java/io/confluent/kafkarest/response/JsonStreamMessageBodyReader.java

```diff
@@ -0,0 +1,71 @@
+/*
+ * Copyright 2022 Confluent Inc.
+ *
+ * Licensed under the Confluent Community License (the "License"); you may not use
+ * this file except in compliance with the License.  You may obtain a copy of the
+ * License at
+ *
+ * http://www.confluent.io/confluent-community-license
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OF ANY KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package io.confluent.kafkarest.response;
+
+import static java.util.Objects.requireNonNull;
+
+import com.fasterxml.jackson.core.JsonParser;
+import com.fasterxml.jackson.databind.JavaType;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import java.io.IOException;
+import java.io.InputStream;
+import java.lang.annotation.Annotation;
+import java.lang.reflect.Type;
+import javax.ws.rs.BadRequestException;
+import javax.ws.rs.WebApplicationException;
+import javax.ws.rs.core.MediaType;
+import javax.ws.rs.core.MultivaluedMap;
+import javax.ws.rs.ext.MessageBodyReader;
+
+/** A {@link MessageBodyReader} for {@link JsonStream}. */
+public final class JsonStreamMessageBodyReader implements MessageBodyReader<JsonStream<?>> {
+  private final ObjectMapper objectMapper;
+
+  public JsonStreamMessageBodyReader(ObjectMapper objectMapper) {
+    this.objectMapper = requireNonNull(objectMapper);
+  }
+
+  @Override
+  public boolean isReadable(
+      Class<?> type, Type genericType, Annotation[] annotations, MediaType mediaType) {
+    return JsonStream.class.equals(type) && MediaType.APPLICATION_JSON_TYPE.equals(mediaType);
+  }
+
+  @Override
+  public JsonStream<?> readFrom(
+      Class<JsonStream<?>> type,
+      Type genericType,
+      Annotation[] annotations,
+      MediaType mediaType,
+      MultivaluedMap<String, String> httpHeaders,
+      InputStream entityStream)
+      throws WebApplicationException {
+    JavaType wrappedType = objectMapper.constructType(genericType).containedType(0);
+    return new JsonStream<>(
```

Comment:
I think I'm missing a concept that's stopping me from reviewing this properly.

How does the returned `() -> objectMapper.readValues` end up being the JsonStream (ie lazy wrapped MappingIterator) that ends up being injected into ProduceAction?

I get that ObjectMapper.readValues returns the right type (based on https://fasterxml.github.io/jackson-databind/javadoc/2.7/com/fasterxml/jackson/databind/ObjectMapper.html#readValues(com.fasterxml.jackson.core.JsonParser,%20java.lang.Class))

But, I don't understand how this JsonStreamMessageBodyReader overwrites the default ObjectMapper that I am theorizing would have previously returned the not-lazy MappingIterator to ProduceAction.

I know it's `config.register(new JsonStreamMessageBodyReader(getJsonMapper()));` but I don't know why.

Can you help me understand please.

Reply:
> But, I don't understand how this JsonStreamMessageBodyReader overwrites the default ObjectMapper that I am theorizing would have previously returned the not-lazy MappingIterator to ProduceAction.

It does not. It specifically implements deserialization for `JsonStream` only, which is not a JSON object, therefore not handled by Jackson. The `isReadable` method tells what this `MessageBodyReader` can deserialize.

Reply:
But how does it know to use this MessageBodyReader, that can deserliaze JsonStream?

The user POSTs to  .../records to produce.

They include some bytes as the data, and the content-type application/json.

The ProduceAction.produce(...) method is called, one of the fields of which is `JsonStream<ProduceRequest> requests)`

How does the code know to inject the field as type JsonStream<ProduceRequests>, I can't see a bind of JsonStream<ProduceRequests> to anything, so I don't understand how the code knows how to make this type of object.  

I can see that the JsonStreamMessageBodyReader returns this type of object when it does a readFrom, which I'm guessing is what populates the `JsonStream<ProduceRequests>` in the ProduceAction.produce() call.  

But I can't see how the code links from the POST to ...../records, and some bytes arriving, into creating a JsonStreamMessageBodyReader (as opposed to some other type of MessageBodyReader).  And how does it know that if a @POST tagged method call contains an arbitrary type of eg `JsonStream<ProduceRequest> requests)` that it needs to use this JsonStreamMessageBodyReader to parse those bytes?

Reply:
Jersey goes over all MessageBodyReaders registered asking isReadable for the type and content-type, at start-up time.